packages/json-rpc-engine/src/v2/JsonRpcServer.ts

```diff
@@ -0,0 +1,193 @@
+import { rpcErrors, serializeError } from '@metamask/rpc-errors';
+import type {
+  Json,
+  JsonRpcId,
+  JsonRpcParams,
+  JsonRpcRequest,
+  JsonRpcResponse,
+  NonEmptyArray,
+} from '@metamask/utils';
+import { hasProperty, isObject } from '@metamask/utils';
+
+import type { JsonRpcMiddleware } from './JsonRpcEngineV2';
+import { JsonRpcEngineV2 } from './JsonRpcEngineV2';
+import type { JsonRpcCall } from './utils';
+import { getUniqueId } from '../getUniqueId';
+
+type HandleError = (error: unknown) => void;
+
+type Options = {
+  handleError: HandleError;
```

Comment:
What is the reason for this being required?

An error handler does seem like a useful feature. I can understand making it required for JSON-RPC notifications at the very least, given that there is nothing else for us to _do_ with the error in that case, it can't be returned.

It does seem like it wouldn't be strictly needed in all cases though, e.g. if the server logged errors as needed in middleware, the top-level handler might not be required.

Reply:
[43f2c94](https://github.com/MetaMask/core/pull/6176/commits/43f2c949c22956eb36b21c248f31d99d21dc7f98)